### ISSUE TITLE
Computing Fst for pairs of populations using popkin 

### ISSUE THREAD

**Issue by privefl** [StoreyLab/popkin#1]
I'm trying to use popkin Fst as an alternative to the W&C one.
I would like to compute the Fst for each pair of populations in the 1000G (26 populations within 5 super-populations).
My code is something like:
```r
kinship <- popkin::popkin(X, subpops = fam2$Population, loci_on_cols = TRUE)
```
where `X` is the matrix of all individuals and a subset of quasi-independent variants (pruning at r2 > 0.2).

Then, I compute the weights
```r
library(dplyr)
weights_tibble  <-
  tibble(superpop = fam2$`Super Population`, pop = fam2$Population) %>% 
  add_count(superpop, name = "n_superpop") %>% 
  add_count(pop, name = "n_pop") %>% 
  mutate(weight = 1 / (n_superpop * n_pop))
```
If I compute `sum(weights_tibble$weight)`, I should get 5 (the number of super populations) according to https://raw.githubusercontent.com/StoreyLab/human-differentiation-manuscript/master/human-origins-01-analysis.pdf.
Which is not the case here.

Then, I compute the Fst for each pair, using `popkin::rescale_popkin()` instead of recomputing the subset kinship.
I then use `popkin::fst()` with `kinship_sub` and `w / sum(w)` (to make sure the weights sum to 1).
```
split_pop <- split(rows_along(fam2), fam2$Population)
npop <- length(split_pop)
dist_fst_popkin <- matrix(0, npop, npop)
colnames(dist_fst_popkin) <- rownames(dist_fst_popkin) <- names(split_pop)
for (i in 2:npop) {
  for (j in 1:(i - 1)) {
    ind <- unlist(split_pop[c(i, j)], use.names = FALSE)
    kinship_sub <- popkin::rescale_popkin(kinship[ind, ind], weights_tibble$pop[ind])
    w <- weights_tibble$weight[ind]
    Fst_popkin <- popkin::fst(kinship_sub, w / sum(w))
    dist_fst_popkin[i, j] <- dist_fst_popkin[j, i] <- print(Fst_popkin)
  }
}
```

**My question is**: would it be the way to do it? Is there anything I'm doing wrong?

**Comment by alexviiia** [comment]
First, none of these are issues with the popkin package (there are no bugs to fix in the package).

Second, the code you shared doesn't match what we used for weights in Human Origins.  In particular, the counts of the superpopulations should be of number of populations, not number of individuals as you have it.  Below is code that does this.

For simplicity, let's rename the names in fam2:
```
names(fam2)[ names(fam2) == 'Super Population'] <- 'superpop'
names(fam2)[ names(fam2) == 'Population'] <- 'pop'
```
Then we create a second tibble where every row is a population instead of an individual:
```
pop <- unique(fam2[, c('superpop', 'pop')])
```
The counts are first for number of individuals per population, then number of populations per superpopulation:
```
fam2 <- add_count(fam2, pop, name = "n_pop") 
pop <- add_count(pop, superpop, name = "n_superpop")
```
You also need the number of unique superpopulations
```
K <- length( unique( pop$superpop ) )
```
The final weights are:
```
weights_tibble <- 1 / ( K * fam2$n_pop * pop$n_superpop[ match( fam2$pop, pop$pop) ] )
```
I confirmed that these weights indeed sum to one.  (They would have summed to 5 if K were excluded, it appears that this is what you intended originally).

Lastly, for computing the pairwise FST matrix, I didn't run it, so I'm not entirely sure that all the steps make sense.  However, I did notice the way you're subsetting the weights is not what I would have done.  The above weights take the double hierarchy of 1000 Genomes into account for computing a global FST.  However, for a parwise FST there isn't a double hierarchy anymore, there are only two populations in each comparison, so the weights should only use the two sample sizes of the populations (so the number of populations per superpopulation is irrelevant here).  For this I would use the function `popkin::weights_subpops`, which returns weights given the labels of the individuals in the subset.  In this case I think this is the right choice (where fam2 is the original, unedited tibble):
```
w <- popkin::weights_subpops( fam2$Population[ ind ] )
```

I just want to close by saying that the spirit of the popkin approach is to not rely on subpopulation labels being meaningful, but instead focusing on the individual-level kinship estimates provided by popkin.  The package has one more function of interest to you, `pwfst`, which estimates an analog of the pairwise FST but between individuals!  I would personally endorse this version of the analysis for individuals, rather than the version you presented for populations, though of course certain analyses require population estimates.  Still, a big point of our analysis of human data is that population labels are sometimes not aligned with genetics (see the 1000 Genomes Hispanics examples on that repository), so any analysis that takes the population partitions as meaningful is inherently flawed.
